server/src/main/java/org/apache/cloudstack/diagnostics/RemoteDiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,144 @@
+        if (hostId == null){
+            s_logger.warn("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+            throw new CloudRuntimeException("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+        }
+
+        final String cmdType = cmd.getType();
+        final String cmdAddress = cmd.getAddress();
+        final String optionalArgunments = cmd.getOptionalArguments();
+
+        String remoteCommand = setupRemoteCommand(cmdType, cmdAddress, optionalArgunments);
+
+        final ExecuteDiagnosticsCommand command = new ExecuteDiagnosticsCommand(remoteCommand);
+        command.setAccessDetail(NetworkElementCommand.ROUTER_IP, routerControlHelper.getRouterControlIp(systemVm.getId()));
+        command.setAccessDetail(NetworkElementCommand.ROUTER_NAME, systemVm.getInstanceName());
+
+        Answer origAnswer;
+        try{
+            origAnswer = agentManager.send(hostId, command);
+        } catch (final OperationTimedoutException e) {
+            s_logger.warn("Timed Out", e);
+            throw new AgentUnavailableException("Unable to send commands to virtual machine ", hostId, e);
+        }
+
+        ExecuteDiagnosticsAnswer answer = null;
+        if (origAnswer instanceof ExecuteDiagnosticsAnswer) {
+            answer = (ExecuteDiagnosticsAnswer) origAnswer;
+        }
+
+        return createRemoteDiagnosisResponse(answer);
+    }
+
+    protected RemoteDiagnosticsResponse createRemoteDiagnosisResponse(ExecuteDiagnosticsAnswer answer){
+        RemoteDiagnosticsResponse response = new RemoteDiagnosticsResponse();
+        response.setResult(answer.getResult());
+        response.setDetails(answer.getDetails());
+        return response;
+    }
+
+    protected String setupRemoteCommand(String cmdType, String cmdAddress, String optionalArguments){
+        String COMMAND_LINE_TEMPLATE = String.format("%s %s", cmdType, cmdAddress);
+        if (optionalArguments != null){
+            final String regex = "^[\\w\\-\\s]+$";
+            final Pattern pattern = Pattern.compile(regex);
+            final boolean hasInvalidChar = pattern.matcher(optionalArguments).find();
+
+            if (!hasInvalidChar){
+                s_logger.error("An Invalid character has been passed as an optional parameter");
+                throw new IllegalArgumentException("Illegal argument passed as optional parameter.");
+            }
+            return String.format("%s %s", COMMAND_LINE_TEMPLATE, optionalArguments);
+        }
+        return COMMAND_LINE_TEMPLATE;
+    }
+
+    @Override
+    public List<Class<?>> getCommands() {
+        List<Class<?>> cmdList = new ArrayList<>();
+        cmdList.add(RemoteDiagnosticsCmd.class);
+        return cmdList;
+    }
+
+    @Override
+    public String getConfigComponentName() {
+        return RemoteDiagnosticsService.class.getSimpleName();
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys() {
```

Comment:
Remove configkeys and relevant interface you've implemented. No need of any global settings, unless you want the global settings as a means to specify optional args (static).

Reply:
Removed use of ConfigKeys